### PR TITLE
Fix 987 (taglist.cpp changes)

### PR DESCRIPTION
**Pull request by clanmills**

### CODE REVIEW

**Comment by D4N** [comment]
Looks good to me @clanmills, could you add a test for this to the test suite?

**Comment by clanmills** [comment]
Yes, I think I will.  That will alert us if/when we make changes to the tags.  I do this later today.

**Comment by clanmills** [comment]
@D4N I'm adding a new test for `taglist` and wondering how to pipe data in the test harness.  I've add taglist to suite.conf. I'd like to run tests such as:
```
$taglist --help | cut -d' ' -f 3-
$taglist ALL | cksum
$taglist all | tail -5
```

I want to filter the output for two reasons:

1) taglist --help echo's the path to taglist which depends on the build machine.  Of course I can modify taglist.cpp to skip that matter.

2) `taglist ALL` produces 5000+ lines of output which is more that 1mb
I can add an option --limit to taglist to restrict the amount of output from the command.

I see there magic in system_tests.py for `post_command_hook`.  Will that pipe stdout and how do I use it?

Another dodge would be to only put the checksum into the test (not stdout).  After all we're only interested in knowing if it changed.  stdout isn't interesting and does not need to be in the test code.

If this causes any pain, I'll modify taglist.cpp to avoid creating work for you.

**Comment by D4N** [comment]
Robin Mills <notifications@github.com> writes:

> @D4N I'm adding a new test for `taglist` and wondering how to pipe data in the test harness.  I've add taglist to suite.conf. I'd like to run tests such as:
> ```
> $taglist --help | cut -d' ' -f 3-
> $taglist ALL | cksum
> $taglist all | tail -5
> ```

This is explicitly not supported to maintain compatibility with Windows
which (by default) lacks all these tools.

>
> I want to filter the output for two reasons:
>
> 1) taglist --help echo's the path to taglist which depends on the build machine.  Of course I can modify taglist.cpp to skip that matter.
>
> 2) `taglist ALL` produces 5000+ lines of output which is more that 1mb
> I can add an option --limit to taglist to restrict the amount of output from the command.
>
> I see there magic in system_tests.py for `post_command_hook`.  Will that pipe stdout and how do I use it?
>

You would want to override `compare_stdout` and implement all these
filters in Python. This is for instance done quite extensively in
`tests/tiff_test/test_tag_compare.py`.

If you need help with porting the shell scripts to python, then I can
help you with that.

**Comment by clanmills** [comment]
Thanks for all those tips, Dan.  I'll do something about this in the next day or two.  Public holiday in England today and the weather's a perfect 30℃.  A day to enjoy the garden and BBQ for dinner.

File: samples/taglist.cpp [hunk 1]
```diff
@@ -14,22 +14,36 @@
 #include <string>
 #include <sstream>
 
+#ifdef   EXV_HAVE_LIBZ
+#include <zlib.h>
```
Comment (D4N) [inline samples/taglist.cpp]: Appveyor chokes on this.

File: samples/taglist.cpp [hunk 2]
```diff
@@ -118,20 +133,37 @@ int main(int argc, char* argv[])
...
     } catch (AnyError& e) {
-        std::cout << "Caught Exiv2 exception '" << e << "'\n";
-        return 1;
+        out << "Caught Exiv2 exception '" << e << "'\n";
+        rc = 1 ;
+    }
+    if ( bChecksum ) {
+#ifdef   EXV_HAVE_LIBZ
+        uLong checksum=0;
+        checksum = adler32(checksum,(const Bytef *)out.str().c_str(),out.str().size());
```
Comment (D4N) [inline samples/taglist.cpp]: If this is only present for testing, then I'd suggest to move this part into the python test suite instead (especially as Python has builtin support for this: https://docs.python.org/3/library/zlib.html#zlib.adler32).

Comment (clanmills) [reply samples/taglist.cpp]: You're right.  Delegating this to the test harness is the correct thing to do.

File: samples/taglist.cpp [hunk 3]
```diff
@@ -58,6 +43,26 @@ int main(int argc, char* argv[])
...
 
+                if (item == "all" || item == "ALL" ) {
+                    const GroupInfo* groupList = ExifTags::groupList();
+                    if (groupList) {
+                        while (groupList->tagList_) {
+                            std::ostringstream tags;
+                            ExifTags::taglist(tags,groupList->groupName_);
+                            // https://en.cppreference.com/w/cpp/string/basic_string/getline
+                            std::istringstream input(tags.str()) ;
+                            for (std::string line,comma(","); std::getline(input, line); ) {
```
Comment (piponazo) [inline samples/taglist.cpp]: [suggestion] Could this be replaced to this version?

```cpp
std::string line;
while (std::getline(input, line)) {
  std::cout << groupList->groupName_ << "."
  << (item == "all" ? line.substr(0,line.find(",")) : line)
  << std::endl;
}
```

I find that line 54 difficult to read.

File: samples/taglist.cpp [hunk 4]
```diff
@@ -120,20 +126,22 @@ int main(int argc, char* argv[])
...
+                << "            Panasonic|Pentax|Sigma|Sony|Iptc|"                                                        << std::endl
+                << "            dc|xmp|xmpRights|xmpMM|xmpBJ|xmpTPg|xmpDM|pdf|photoshop|crs|tiff|exif|aux|iptc|all|ALL"  << std::endl
+                << "           ]"                                      << std::endl
+                << "Print Exif tags, MakerNote tags, or Iptc datasets" << std::endl
+                ;
         }
-        return rc;
     } catch (AnyError& e) {
         std::cout << "Caught Exiv2 exception '" << e << "'\n";
-        return 1;
```
Comment (piponazo) [inline samples/taglist.cpp]: Could we replace the `rc` values with the standard `EXIT_SUCCESS` and `EXIT_FAILURE` values ?

**Review by piponazo — DISMISSED** [review]
LGTM! 

I am going to push to commits with the changes I proposed. We can always drop them if you do not agree with them.

**Review by clanmills — COMMENTED** [review]
I totally agree.  Thanks for doing this.

**Review by D4N — DISMISSED** [review]
I've rebased, reordered and squashed some commits (but kept the overall diff the same). I've found a little mistake though: the man page got accidentally duplicated. I've removed this duplication.

Looks good to me now and I'd merge this once the CI is green. Thanks for participating @clanmills & @piponazo

**Comment by D4N** [comment]
Well damn, now the tests are failing… And I've just verified it again: there are **no** code changes introduced by my rebase.

I have the suspicion though that the failing tests are caused by merging #988, which added new tests. Maybe we should actually refactor the tests, because currently it is quite hard to find out what they are actually doing.

**Comment by clanmills** [comment]
I'll deal with the tests failing, Dan.  Not sure what's happened, the line-count is wrong for sure.

**Comment by clanmills** [comment]
Subtle.  #988 adds new tags for Canon AutoFocus.  So, test_issue_987 correct detects this.  Incidentally, I've written a little script `adler32` to generate the output required in the test script:

```
#!/usr/bin/env python3
import zlib
import sys
data = sys.stdin.read()
print('%d,%d' % ( len(data.split('\n')), zlib.adler32(bytes(data,'utf-8') )))
```
And in action:
```
.../build $ for a in --help all ALL Groups nothing; do bin/taglist $a 2>/dev/null| adler32 ; done
8,198867862
5426,1072295739
5426,838787056
107,430791180
8,198867862
593 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ 
```

**Comment by clanmills** [comment]
@D4N Can you give ci/gitlab/fix_987 a kick in the back-side.  He's failing on CentOS say "python3 not found".   Everybody else is happy.  I thought there was a button somewhere to say "re-run the CI".

**Review by tester0077 — APPROVED** [review]

**Pull request merged**